hare/hare_test.go

```diff
@@ -98,6 +94,11 @@ func noopPubSub(tb testing.TB) pubsub.PublishSubsciber {
 	return publisher
 }
 
+func newMockMesh(tb testing.TB) *mocks.Mockmesh {
+	tb.Helper()
+	return mocks.NewMockmesh(gomock.NewController(tb))
```

Comment:
I think `gomock.NewController` should only be called once per test. So it would be better to pass in the controller here, but at that point we can just completely inline this function.

Reply:
what's the reason for such restriction tho?

Reply:
gomock isn't super clear in its documentation:

> A Controller represents the top-level control of a mock ecosystem. It defines the scope and lifetime of mock objects, as well as their expectations. It is safe to call Controller's methods from multiple goroutines. Each test should create a new Controller and invoke Finish via defer.

It's also outdated, `Finish` doesn't need to be called at all any more since the introduction of `t.Cleanup`.

I never tried to use multiple controllers so idk what happens in that case. I assumed then possibly missing or unexpected calls would be reported twice?

Reply:
But thanks for making me look it up ❤️ so I saw this: 

https://pkg.go.dev/github.com/golang/mock/gomock#WithContext

I did not know that this exists, this should be super helpful in tests doing concurrency with mocks (otherwise they often get stuck). Maybe this also helps you with `Test_multipleCPs` @piersy ?

Reply:
1. the reason i am resisting here is that the hare test setup is convoluted. if i tried to make the same ctrl that will be a huge change.
2. i think the idea of not having to call `Finish` explicitly implies that gomock is tracking the ctrl underneath the cover. in any case, the tests failed if i don't set expectations so it definitely is checking....

Reply:
`NewController` calls `t.Cleanup` with `Finish` if it is supported: https://github.com/golang/mock/blob/aba2ff9a6844d5e3289e8472d3217d5b3090f083/gomock/controller.go#L136-L141

But yeah, looks like its safe to have multiple controllers... 👍